### ISSUE TITLE
Feature Request - add key event

### ISSUE THREAD

**Issue by j4587698** [xamarin/Xamarin.Forms#3444]
### Description

In XF3.0, most support platforms have physical keyboards like mac wpf gtk# and uwp. Android can also be connected to keyboard. So I need keydown keyup or keypress event, not only back button.
I need to monitor the shortcut keys like ctrl+A F1 etc.

**Comment by AdhamAliAhmed** [comment]
Why not Android And iOS even if they are not connected to a physical keyboards?

**Comment by brettclutch** [comment]
I'm trying to do the same thing in a CustomRenderer, I hook up the KeyDown event like so but I recieve no call to the event despite trying tricks with focus to get the CustomRenderer to have focus. I feel like I'm missing something or its broke. 

 protected override void OnElementChanged(ElementChangedEventArgs<TFormsView> e)
		{
			if (e.OldElement != null)
			{
				var oldController = (IFlowSKCanvasViewController)e.OldElement;
				// unsubscribe from events
				oldController.SurfaceInvalidated -= OnSurfaceInvalidated;
				oldController.GetCanvasSize -= OnGetCanvasSize;
			}

			if (e.NewElement != null)
			{
				var newController = (IFlowSKCanvasViewController)e.NewElement;

				// create the native view
				if (Control == null)
				{
					var view = CreateNativeControl();
					view.PaintSurface += OnPaintSurface;
					SetNativeControl(view);
				}

                Control.KeyDown += Control_KeyDown;
}

Here Control_KeyDown if a breakpoint is inputted will never fire.

**Comment by stevehurcombe** [comment]
+1 here. I think this would be an important addition to any cross platform dev library!

Personally I want to detect a Ctrl-S (for Save) combination when a Bluetooth keyboard is attached to a mobile device.

**Comment by txarito** [comment]
+1.
I can't understand how OnKeyDown method isn't accesible from Xamarin.Form ContentPage.

**Comment by jwosty** [comment]
I really want this, too.

**Comment by delphian** [comment]
+1, Yes please. For the UWP version of my application, or an Android with keyboard, it would be great to detect keystrokes from the ContentPage.

**Comment by charlesroddie** [comment]
Presumably the right way to do this is to allow a listener to be attached to any `View`, similarly to touch?

**Comment by jwosty** [comment]
It would also be nice if gesture event handlers also told you which keyboard buttons were pressed at the time of the gesture. You could use this to cause a click to behave differently from a CTRL+click. However, if key pressed events (or keyup+keydown) were implemented, you could use that to do it yourself.

**Comment by volcano619** [comment]
I'm stuck at not able to get a event when an entry is empty, and delete is pressed, i hope we can add keyboard events to views

**Comment by ArtjomP** [comment]
Let me add my 2 cents.
How about to have such a thing as KeyGestureRecognizer?

**Comment by volcano619** [comment]
are we still considering this @samhouts ?

**Comment by samhouts** [comment]
Yes, we are still considering this. What would the Xamarin.Forms API look like? Can you show examples of how to use it with C# and XAML? Thanks!

**Comment by ArtjomP** [comment]
OnKeyUp (Down) event for ContentPage would be ok.
I would like more something like this one:
```xaml
<Grid>
      <Grid.GestureRecognizers>
          <KeyboardGestureRecognizer
                  Key="X"
                  AndAlt="True"
                  AndCtrl="False"
                  AndShift="True"
                  Command="{Binding MyCommand}"
                  Event="..."/>
      </Grid.GestureRecognizers>
</Grid>

**Comment by kkentt** [comment]
@samhouts Anything about this ?

**Comment by GPManuel** [comment]
> OnKeyUp (Down) event for ContentPage would be ok.
> I would like more something like this one:
> 
> ```
> <Grid>
>       <Grid.GestureRecognizers>
>           <KeyboardGestureRecognizer
>                   Key="X"
>                   AndAlt="True"
>                   AndCtrl="False"
>                   AndShift="True"
>                   Command="{Binding MyCommand}"
>                   Event="..."/>
>       </Grid.GestureRecognizers>
> </Grid>
> ```

does it exist?

**Comment by WilliamWatterson86** [comment]
Has anyone been able to get a command to fire from a keyboard?  Eg. I have a frame and some buttons with tap gesture and really need this to be triggered with the space or enter buttons.

**Comment by adopilot** [comment]
What about approcuh https://forums.xamarin.com/discussion/80116/how-can-i-detect-key-events-in-xamarin-forms described by ClayBrinlee. How bad is in start of app for every single keydown call MessagingCenter.Send ? I do subscribe only in one place but does this still slowing down app if there is no subscribers in MessagingCenter ?

**Comment by volcano619** [comment]
> Has anyone been able to get a command to fire from a keyboard? Eg. I have a frame and some buttons with tap gesture and really need this to be triggered with the space or enter buttons.

you might be able to do this in renderer i believe, i think android/ios natively do give that support, for the time being we might be able to use this
